server/websocket.go

```diff
@@ -1394,9 +1395,11 @@ func (c *client) wsCollapsePtoNB() (net.Buffers, int64) {
 					if endStart {
 						fhIdx = startFrame()
 					}
-					bufs = append(bufs, b[:total])
+					buf := nbPoolGet(total)
```

Comment:
I need some help understanding this part. We grab a pool buf and then copy a slice of b into only to copy it immediately again into bufs? Why?

Reply:
Every slice that gets put into `bufs` and returned eventually gets casted back to an array type at the full capacity and put back into the NB pool. This keeps allocations down compared to pooling slice references themselves and ends up being faster. 

If you chop the slice from the front to slide the window forward and then append those, multiple references to the same underlying array can get returned to the pool, which creates races on the underlying arrays. 

Getting fresh arrays from the pool for each limited segment avoids this problem as it guarantees only one reference per underlying array. It's extremely subtle hence the comments added on the NB pool.

Reply:
Feels strange that we copy `b[:total]` twice in that line, feels inefficient no?

Reply:
It's not two copies of `b[total:]`, it's one from `b[total:]` to `buf` from the pool and then adds `buf` into `bufs` which is returned.

Reply:
Yes correct, apologies. We copy the low level buffer once and then add it to the list of bufs.